### PR TITLE
ctpqc:fixes

### PR DESCRIPTION
**Pull request by lietava**
TH1FRatio used.

### CODE REVIEW

File: Modules/CTP/src/RawDataReaderCheck.cxx [hunk 1]
```diff
@@ -99,6 +99,8 @@ Quality RawDataReaderCheck::check(std::map<std::string, std::shared_ptr<MonitorO
   if (lhcifdata == nullptr) {
     ILOG(Info, Support) << "LHC data not found for timestamp:" << mTimestamp << ENDM;
     return result;
+  } else {
+    ILOG(Info, Support) << "LHC data found for timestamp:" << mTimestamp << ENDM;
```
Comment (Barthelemy) [inline Modules/CTP/src/RawDataReaderCheck.cxx:103]: This message will be printed every time the check is called. Is it what you want ? shouldn't it be a debug message ?

File: Modules/CTP/src/RawDataReaderCheck.cxx [hunk 2]
```diff
@@ -142,7 +143,7 @@ Quality RawDataReaderCheck::check(std::map<std::string, std::shared_ptr<MonitorO
     } else {
       auto* h = dynamic_cast<TH1F*>(mo->getObject());
       if (!h) {
-        ILOG(Info, Support) << "histogram for input/class rate is not found for check" << ENDM;
+        // ILOG(Debug, Devel) << "histogram for input/class rate is not found for check" << ENDM;
```
Comment (Barthelemy) [inline Modules/CTP/src/RawDataReaderCheck.cxx:146]: We advise not to comment lines but to simply remove them. Git will remember.

File: Modules/CTP/src/RawDataReaderCheck.cxx [hunk 3]
```diff
@@ -161,10 +162,10 @@ Quality RawDataReaderCheck::check(std::map<std::string, std::shared_ptr<MonitorO
           h->Scale(1. / h->GetBinContent(mIndexMBclass));
         }
       }
-
       if (cycleCounter > 1) { // skipping the check for first two cycles, as the first one is arbitrary long
         TH1F* fHistDifference = (TH1F*)h->Clone();
-        TH1F* fHistPrev;
+        TH1F* fHistPrev = nullptr;
+        ;
```
Comment (Barthelemy) [inline Modules/CTP/src/RawDataReaderCheck.cxx:168]: unneeded `;`

File: Modules/CTP/src/RawDataReaderCheck.cxx [hunk 4]
```diff
@@ -199,7 +200,12 @@ Quality RawDataReaderCheck::check(std::map<std::string, std::shared_ptr<MonitorO
         }
         if (!relativeRates) {
           fHistDifference->Scale(1. / (mCycleDuration * mFraction));
-          fHistPrev->Scale(1. / (mCycleDuration * mFraction));
+          if (fHistPrev == nullptr) {
+            std::cout << " crashing here" << std::endl;
```
Comment (Barthelemy) [inline Modules/CTP/src/RawDataReaderCheck.cxx:204]: Please use `ILOG` and not `cout`. Also, I am not sure this message should really go to production. A more informative message would be better.

**Review by Barthelemy — CHANGES_REQUESTED** [review]
Thank you for your PR. 
Could you please set a title to the PR that is less general and gives and idea of what the changes are about ? 
I have noted a couple of little things to change, but not much. I let you fix them and I will then merge the PR.

**Comment by lietava** [comment]
Hi @Barthelemy this is still work in progress.  Thanks for your comments - I will use them. When ready I will click Ready for review.

**Comment by Barthelemy** [comment]
@lietava Ok, understood. You can use the prefix `[WIP]` in the title of your PR to signify that we should not review yet.

**Comment by lietava** [comment]
@Barthelemy : did not know about WIP - thanks.

**Comment by lietava** [comment]
@Barthelemy , @lhusova , @aferrero2707 please, have a look.

File: Modules/CTP/src/RawDataReaderCheck.cxx (hunk 1 above)
Comment (lietava) [reply Modules/CTP/src/RawDataReaderCheck.cxx]: outdated

**Review by aferrero2707 — COMMENTED** [review]
As a general remark, I would suggest to use TH1FRatio objects in order to normalize the histograms. This would have a number of advantages:
* you could ten use the moving windows in order to generate versions of the plots that only contain data from the last QC cycle
* you could use the moving windows output as input for the trending task, such that each point in the graph would only contain data from the last cycle
* you could avoid the normalization gymnastics in the checker
* the TH1FRatio objects would allow to keep the information on the number of entries for each bin, because they internally contain a copy of the not-normalized histogram

Later in the week I can eventually make a PR on your [lietava:ctpqc](https://github.com/lietava/QualityControl/tree/ctpqc) branch with the required changes, if you wish.

**Comment by Barthelemy** [comment]
Given the extensive review made by Andrea, I will not review this PR. Once it is merged, I will then proceed with the general CTP QC review.

**Comment by lietava** [comment]
@aferrero2707 : please, have a look.

**Comment by Barthelemy** [comment]
@lietava it seems that you have a conflict

**Comment by Barthelemy** [comment]
@aferrero2707 Could you have a last look ?

**Comment by aferrero2707** [comment]
@Barthelemy yes, absolutely. I was doing some tests of the PR locally, but my development machine went down due to some electrical intervention. Now it is up again, I will finish the check by tomorrow.

File: Modules/CTP/src/RawDataReaderCheck.cxx [hunk 5]
```diff
@@ -91,154 +87,66 @@ void RawDataReaderCheck::configure()
...
         delete fHistInputPrevious;
-        fHistInputPrevious = (TH1F*)h->Clone();
       }
-      if (inputRates && relativeRates) {
+      fHistInputPrevious = (TH1F*)h->Clone();
+    } else if (mo->getName() == "inputRatio") {
+      if (fHistInputRatioPrevious) {
         delete fHistInputRatioPrevious;
-        fHistInputRatioPrevious = (TH1F*)h->Clone();
       }
```
Comment (aferrero2707) [inline Modules/CTP/src/RawDataReaderCheck.cxx:135]: @lietava It seems to me that `inputRatio` is not checked, and also the re-scaling with respec to the MB trigger (if I understand correctly the purpose of the ratio plots) is not applied in the task. Can you check?

File: Modules/CTP/src/RawDataReaderCheck.cxx [hunk 6]
```diff
@@ -91,154 +87,66 @@ void RawDataReaderCheck::configure()
...
+      fHistClassesPrevious = (TH1F*)h->Clone();
+    } else if (mo->getName() == "classRatio") {
+      if (fHistClassRatioPrevious) {
         delete fHistClassRatioPrevious;
-        fHistClassRatioPrevious = (TH1F*)h->Clone();
-      }
-      if (!relativeRates && inputRates) {
-        cycleCounter++;
       }
+      fHistClassRatioPrevious = (TH1F*)h->Clone();
```
Comment (aferrero2707) [inline Modules/CTP/src/RawDataReaderCheck.cxx:147]: Same as previous comment

**Review by aferrero2707 — APPROVED** [review]
I left a couple of minor comments, but overall it is ok for me.

**Comment by aferrero2707** [comment]
@Barthelemy @lietava I left a couple of additional comments, but overall it is ok for me. I propose to merge the code, and most likely re-iterate following the feedback in production.

The current code should already allow to add CTP into the shifters QC layouts and have it checked regularly.

**Pull request merged**